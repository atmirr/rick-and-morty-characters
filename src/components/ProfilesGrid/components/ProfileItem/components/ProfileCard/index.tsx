import React from 'react';
import styled from 'styled-components';
import BasicInfoItem from './components/BasicInfoItem';
import DetailItem from './components/DetailItem';
import LocationItem from './components/LocationItem';
import EpisodesItem from './components/EpisodesItem';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: #ffffff;
  width: 230px;
  text-align: center;
  padding: 1.3em 1.3em 0.5em 1.3em;
  margin: 1.1em 0.9em;
  border-radius: 15px;
  box-shadow: 0 0 6px 4px #e3e3e3;
  transition: all 300ms ease-out;
  &:hover {
    box-shadow: 0 0 14px 2px #aaaaaa;
  }
`;

const ImageWrapper = styled.div`
  display: flex;
  cursor: pointer;
  border-radius: 15px;
  & img {
    border-radius: 15px;
  }
`;
const DetailsWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  text-align: left;
  padding: 12px 0;
`;

const Name = styled.span`
  font-size: 22px;
  padding-bottom: 2px;
`;

const LocationItems = styled.div`
  padding: 2px 0;
`;

type PropTypes = {
  name: string;
  species: string;
  status: string;
  gender: string;
  image: string;
  episodes: string[];
  locationName?: string;
  locationType?: string;
  dimension?: string;
  numberOfResidents?: number;
};

function ProfileCard({
  name,
  species,
  status,
  gender,
  image,
  episodes,
  locationName,
  locationType,
  dimension,
  numberOfResidents,
}: PropTypes): React.ReactElement {
  return (
    <Container data-testid="profile-card">
      <ImageWrapper>
        <img src={image} alt={name} width={220} height={220} />
      </ImageWrapper>
      <DetailsWrapper>
        <Name>{name}</Name>
        <BasicInfoItem species={species} status={status} gender={gender} />
        <LocationItems>
          <LocationItem label="Dimension">{dimension}</LocationItem>
          <LocationItem label="Location">{locationName}</LocationItem>
          <LocationItem label="Location Type">{locationType}</LocationItem>
        </LocationItems>
        <DetailItem label="Number of Residents">{numberOfResidents}</DetailItem>
        {episodes && <EpisodesItem>{episodes}</EpisodesItem>}
      </DetailsWrapper>
    </Container>
  );
}

export default ProfileCard;
