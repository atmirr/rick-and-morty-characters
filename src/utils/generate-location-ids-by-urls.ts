export const generateLocationIdsByUrls = (urls: string[]): string[] =>
  urls.reduce((result: string[], url: string) => {
    const splittedUrl = url.split('/');
    const id = splittedUrl[splittedUrl.length - 1];
    return id ? [...result, id] : [...result];
  }, []);
