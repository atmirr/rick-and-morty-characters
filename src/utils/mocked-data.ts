export const characters = [
  {
    id: 1,
    name: 'Rick Sanchez',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:48:46.250Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/1',
  },
  {
    id: 2,
    name: 'Morty Smith',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:50:21.651Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
      'https://rickandmortyapi.com/api/episode/3',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/2',
  },
];

export const episodes = [
  'Rixty Minutes',
  'Something Ricked This Way Comes',
  'Close Rick-counters of the Rick Kind',
  'Ricksy Business',
  'A Rickle in Time',
];

export const locations = [
  {
    name: 'Citadel of Ricks',
    type: 'Space station',
    dimension: 'unknown',
  },
  {
    name: 'Earth (Replacement Dimension)',
    type: 'Planet',
    dimension: 'Replacement Dimension',
  },
];
