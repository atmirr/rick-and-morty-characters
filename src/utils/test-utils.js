import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockedInitialState = {
  characters: { items: [], nextPage: null, error: null, isLoading: false },
  episodes: { items: [], nextPage: null, error: null, isLoading: false },
  locations: { items: [], error: null, isLoading: false },
};

function render(
  ui,
  {
    initialState = mockedInitialState,
    store = configureStore()(initialState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
