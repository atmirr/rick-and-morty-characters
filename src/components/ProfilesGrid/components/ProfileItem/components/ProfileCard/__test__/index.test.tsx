import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { screen, render } from 'src/utils/test-utils';
import { characters, episodes, locations } from 'src/utils/mocked-data';
import ProfileCard from '../index';

const props = {
  name: characters[0].name,
  species: characters[0].species,
  status: characters[0].status,
  gender: characters[0].gender,
  image: characters[0].image,
  episodes,
  locationName: locations[0].name,
  locationType: locations[0].type,
  dimension: locations[0].dimension,
  numberOfResidents: 10,
};

describe('<ProfileCard />', () => {
  it('should renders the component', () => {
    const { container } = render(<ProfileCard {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display profile details when all the props has been passed', () => {
    const { container } = render(<ProfileCard {...props} />);
    expect(screen.queryAllByTestId('profile-card').length).toBe(1);
    const imageNodes = container.querySelectorAll('img');
    expect(imageNodes.length).toBe(1);
    expect(screen.queryAllByText(/Rick Sanchez/i).length).toBe(1);
    expect(screen.queryAllByText(/Human/i).length).toBe(1);
    expect(screen.queryAllByText(/(Male)/i).length).toBe(1);
    expect(screen.queryAllByText(/Alive/i).length).toBe(1);
    expect(screen.queryAllByTestId('episodes-item').length).toBe(1);
    expect(screen.queryAllByText(/Citadel of Ricks/i).length).toBe(1);
    expect(screen.queryAllByText(/Space station/i).length).toBe(1);
    expect(screen.queryAllByText(/unknown/i).length).toBe(1);
    expect(screen.queryAllByText(/10/i).length).toBe(1);
  });
});
