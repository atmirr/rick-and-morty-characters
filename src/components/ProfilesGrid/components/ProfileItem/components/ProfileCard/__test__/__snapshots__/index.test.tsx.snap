// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileCard /> should renders the component 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 16px;
  line-height: 1.4em;
  padding-bottom: 2px;
  color: #797979;
}

.c8 {
  padding: 0 3px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  text-transform: capitalize;
}

.c6 {
  height: 7px;
  width: 7px;
  margin-right: 4px;
  background-color: #55cc44;
  border-radius: 50%;
}

.c9 {
  text-transform: capitalize;
}

.c10 {
  padding-left: 3px;
  font-size: 13px;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 6px 1px;
}

.c16 {
  font-size: 16px;
  font-weight: 500;
  color: #7c7c7c;
}

.c17 {
  padding-left: 3px;
  color: #222222;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 4px 0;
}

.c13 {
  font-size: 15px;
  color: #797979;
}

.c14 {
  font-size: 16px;
}

.c18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 6px 1px;
}

.c19 {
  font-size: 16px;
  font-weight: 500;
  color: #7c7c7c;
}

.c20 {
  padding-left: 3px;
  color: #222222;
}

.c21 {
  font-weight: 500;
  color: #1976d2;
  cursor: pointer;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #ffffff;
  width: 230px;
  text-align: center;
  padding: 1.3em 1.3em 0.5em 1.3em;
  margin: 1.1em 0.9em;
  border-radius: 15px;
  box-shadow: 0 0 6px 4px #e3e3e3;
  -webkit-transition: all 300ms ease-out;
  transition: all 300ms ease-out;
}

.c0:hover {
  box-shadow: 0 0 14px 2px #aaaaaa;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  border-radius: 15px;
}

.c1 img {
  border-radius: 15px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: left;
  padding: 12px 0;
}

.c3 {
  font-size: 22px;
  padding-bottom: 2px;
}

.c11 {
  padding: 2px 0;
}

<div
  class="c0"
  data-testid="profile-card"
>
  <div
    class="c1"
  >
    <img
      alt="Rick Sanchez"
      height="220"
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      width="220"
    />
  </div>
  <div
    class="c2"
  >
    <span
      class="c3"
    >
      Rick Sanchez
    </span>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <span
          class="c6"
          data-testid="status-icon"
        />
        <span
          class="c7"
        >
          Alive
        </span>
      </div>
      <span
        class="c8"
      >
        -
      </span>
      <div
        class="c5"
      >
        <span
          class="c9"
        >
          Human
        </span>
        <span
          class="c10"
        >
          (
          Male
          )
        </span>
      </div>
    </div>
    <div
      class="c11"
    >
      <div
        class="c12"
      >
        <span
          class="c13"
        >
          Dimension
          :
        </span>
        <span
          class="c14"
        >
          unknown
        </span>
      </div>
      <div
        class="c12"
      >
        <span
          class="c13"
        >
          Location
          :
        </span>
        <span
          class="c14"
        >
          Citadel of Ricks
        </span>
      </div>
      <div
        class="c12"
      >
        <span
          class="c13"
        >
          Location Type
          :
        </span>
        <span
          class="c14"
        >
          Space station
        </span>
      </div>
    </div>
    <div
      class="c15"
    >
      <span
        class="c16"
      >
        Number of Residents
        :
      </span>
      <span
        class="c17"
      >
        10
      </span>
    </div>
    <div
      class="c18"
      data-testid="episodes-item"
    >
      <span
        class="c19"
      >
        Episodes
        :
      </span>
      <div>
        <span
          class="c20"
        >
          Rixty Minutes
        </span>
         
        <a
          class="c21"
        >
          and 4 more
        </a>
      </div>
    </div>
  </div>
</div>
`;
